### ISSUE TITLE
Hard to tell when the final note is actually over

### ISSUE THREAD

**Issue by kavika13** [schellingb/TinySoundFont#27]
I'm not sure if I'm doing things incorrectly, but `tml_get_info` seems to return the begin timestamp of the last note in `out_time_length`, and not include the lingering time. If I stop playback at exactly this time, then the final note gets cut off in the particular song I'm playing. I'm also not sure if it's cutting off on-tempo or not.

Is there some mechanism I've missed, or work-around that I can do? There are a couple I thought of ("just wait a bit longer" and "double-output so I can sample the volume"), but they don't seem ideal. I am planning on specifying a mid-song loop point, running to the end, and looping back to that point, so I think I need to get that last note to be the correct length for the tempo.

I might be able to get enough info if I had a way of determining "final" tempo, and got info from the midi header, but as far as I can tell, those are buried inside the implementation right now.

**Comment by kavika13** [comment]
I figured out that when the current message is null, I can query the soundfont voices to see which voices have a value `playingPreset != -1` and `ampenv.segment < TSF_SEGMENT_DONE || ampenv.parameters.release`. Once that is false for all of them, they seem to be finished playing.

I found that out because it these sorts of checks are done in a number of places to verify that things are playing before shutting them off (`tsf_voice_render`, `tsf_channel_sounds_off_all`, etc).

Maybe this should be a point of documentation, or reflected in the demo app? Or maybe a simple function or two would be useful here; `tsf_is_voice_playing` and maybe `tsf_are_any_voices_playing`.

**Comment by adrianmcroft** [comment]
Hi kavika, I'm looking for something similar for my note-stealing. I popped this routine temporarily into TSF - I'm pretty sure that's where you got to.... 

```
int count_playing_notes(tsf* f)
{
	int playingCount = 0;

	struct tsf_voice *v = f->voices, *vEnd = v + f->voiceNum;
	for (; v != vEnd; v++)
	{
		if ((v->playingPreset >= 0) && (v->playingKey != 0)) playingCount++;
	}

	return playingCount;
}

**Comment by schellingb** [comment]
Yeah, `tml_get_info` just returns the timestamp of the last message in `out_time_length`. So if the MIDI ends with just NOTE_ON without a matching NOTE_OFF or any other final message (i.e. ALL_SOUND_OFF) then it is quite misleading. Depending on what bank/instrument is active for the final note(s) even with NOTE_OFF tml would have no idea how long the instruments release time will be.

I agree that the function you suggested or maybe a `tsf_active_voice_count` would be good to have.
But still a MIDI player that wants to absolutely correctly report a song duration to the user ahead of playing would be hard. Hmm, I wonder what other players do.

**Comment by kavika13** [comment]
> Yeah, `tml_get_info` just returns the timestamp of the last message in `out_time_length`. So if the MIDI ends with just NOTE_ON without a matching NOTE_OFF or any other final message (i.e. ALL_SOUND_OFF) then it is quite misleading. Depending on what bank/instrument is active for the final note(s) even with NOTE_OFF tml would have no idea how long the instruments release time will be.
> 
> I agree that the function you suggested or maybe a `tsf_active_voice_count` would be good to have.
> But still a MIDI player that wants to absolutely correctly report a song duration to the user ahead of playing would be hard. Hmm, I wonder what other players do.

I know I blathered on in the beginning of this ticket, sorry about that! I think this can be split into two tickets.

For this ticket, I think the right solution for a user is:
1. Detect end of message stream (`->next == null`)
2. Call `tsf_channel_note_off_all` (if an implementation wants to wait a bit first, they can. If the song has already turned all the notes off then there's no reason to wait. If it hasn't turned them all off, then it's not super important when they decide to send this. It would probably be best if they tempo match, and that could be enabled by #28 ).
3. Detect when all the voices have extinguished, then unhook the sound sample inserter callbacks. This is the part I think needs additional support, via this ticket, either with the function(s) I proposed, or a simple active voice count function.
